### PR TITLE
zed load: add fancy stats updates

### LINKED ISSUE #3354: zed load: progress bar
Add progress bar for zed load provided the current shell is a tty.

### PR DESCRIPTION
**Pull request by mattnibs**
Add stats updates to zed load if -q is not enabled and the current shell
is a tty.

Closes #3354

### CODE REVIEW

File: cli/inputflags/flags.go [hunk 1]
```diff
@@ -66,3 +69,90 @@ func (f *Flags) Open(ctx context.Context, zctx *zed.Context, engine storage.Engi
...
+			}
+			fmt.Fprintln(os.Stderr, err)
+			continue
+		}
+		readers = append(readers, file)
+	}
+	return readers, nil
+}
+
+func openFile(ctx context.Context, zctx *zed.Context, engine storage.Engine, path string, opts anyio.ReaderOpts) (*StatsReader, error) {
```
Comment (mattnibs) [inline cli/inputflags/flags.go:93]: It was kind of gross having to replicate the logic of anyio.Open here but I need to wrap the Readers to count the bytes read- didn't see another way. Open to suggestions.

File: pkg/display/display.go [hunk 2]
```diff
@@ -61,7 +65,11 @@ func (d *Display) Bypass() io.Writer {
 }
 
 func (d *Display) Close() {
-	close(d.close)
+	select {
+	case <-d.close:
+	default:
+		close(d.close)
+	}
```
Comment (nwt) [inline pkg/display/display.go:72]: I think this has a race condition. If you enter the default case but another goroutine closes `d.close` before you do, you'll panic.

Comment (mattnibs) [reply pkg/display/display.go]: I'll trade you a close for a ctx, cancel?

File: cli/inputflags/flags.go (hunk 1 above)
Comment (nwt) [reply cli/inputflags/flags.go:93]: Create a storage.Engine wrapper with a Get method that, in turn, wraps a storage.Reader with something like your StatsReader before returning it and then provides a way to get stats for the wrapped readers!

Comment (mattnibs) [reply cli/inputflags/flags.go]: Great idea!

File: cmd/zed/load/command.go [hunk 3]
```diff
@@ -92,3 +117,72 @@ func (c *Command) Run(args []string) error {
...
+	e.readers = append(e.readers, counter)
+	return counter, nil
+}
+
+func (e *engineWrap) status() (units.Bytes, units.Bytes, int) {
+	var read units.Bytes
+	for _, r := range e.readers {
+		read += units.Bytes(r.bytesRead())
+	}
+	return read, e.bytesTotal, int(atomic.LoadInt32(&e.completed))
```
Comment (nwt) [inline cmd/zed/load/command.go:168]:
```suggestion
	var read int64
	for _, r := range e.readers {
		read += r.bytesRead()
	}
	return units.Bytes(read), e.bytesTotal, int(atomic.LoadInt32(&e.completed))
```

**Review by nwt — APPROVED** [review]

File: pkg/display/display.go [hunk 4]
```diff
@@ -18,17 +19,24 @@ type Display struct {
 	interval time.Duration
 	updater  Displayer
 	buffer   *bytes.Buffer
-	close    chan struct{}
+	ctx      context.Context
+	cancel   context.CancelFunc
```
[the remark this answers is not in the document]
Comment (mattnibs) [reply pkg/display/display.go:23]: Agreed, good idea.

**Pull request merged**